include/minstrobe.hpp

```diff
@@ -0,0 +1,516 @@
+    std::deque<value_t> window_values{};
+
+    //!\brief The number of values in one window.
+    size_t window_size{};
+
+    //!\brief Increments iterator by 1.
+    void next_unique_minstrobe()
+    {
+        while (!next_minstrobe()) {}
+    }
+
+    //!\brief Returns new window value of the first iterator.
+    auto window_value() const
+    {
+        return *urng1_iterator;
+    }
+
+    //!\brief Returns new window value of the first iterator.
+    auto first_window_value() const
+    {
+        return *first_iterator;
+    }
+
+    //!\brief Advances the window of the first iterator to the next position.
+    void advance_window()
+    {
+        ++urng1_iterator;
+    }
+
+    //!\brief Advances the window of the first iterator to the next position.
+    void advance_all_windows()
+    {
+        ++first_iterator;
+	++urng1_iterator;
+    }
+
+    //!\brief Calculates minstrobes for the first window.
+    void window_first(const size_t window_min, const size_t window_max)
+    {
+	window_size = (window_max - window_min + 1);
+
+	if (window_size == 0u)
+            return;
+
+	first_iterator = urng1_iterator;
+	std::advance(urng1_iterator, window_min);
+
+        for (int i = 1u; i < window_size; ++i)
+        {
+            window_values.push_back(window_value());
+            advance_window();
+        }
+        window_values.push_back(window_value());
+
+        auto minstrobe_it = std::ranges::min_element(window_values, std::less_equal<value_t>{});
+        minstrobe_value = std::make_tuple(first_window_value(), *minstrobe_it);
+        minstrobe_position_offset = std::distance(std::begin(window_values), minstrobe_it);
+
+    }
+
+    /*!\brief Calculates the next minstrobe value.
+     * \returns True, if new minstrobe is found or end is reached. Otherwise returns false.
+     * \details
+     * For the following windows, we remove the first window value (is now not in window_values) and add the new
+     * value that results from the window shifting.
+     */
+    bool next_minstrobe()
+    {
+    	advance_all_windows();
+
+        if (urng1_iterator == urng1_sentinel)
+            return true;
+
+        value_t const new_value = first_window_value();
+        value_t const sw_new_value = window_value();
+
+
+        std::get<0>(minstrobe_value) = new_value;
+
+        window_values.pop_front();
+        window_values.push_back(sw_new_value);
+
+	if (minstrobe_position_offset == 0)
+	{
+		auto minstrobe_it = std::ranges::min_element(window_values, std::less_equal<value_t>{});
+		std::get<1>(minstrobe_value) = *minstrobe_it;
+		minstrobe_position_offset = std::distance(std::begin(window_values), minstrobe_it);
+		return true;
+
+	}
+
+	if (sw_new_value < std::get<1>(minstrobe_value))
+	     {
+	          std::get<1>(minstrobe_value) = sw_new_value;
+	          minstrobe_position_offset = window_values.size() - 1;
+	          return true;
+	     }
+
+	--minstrobe_position_offset;
+	return true;
+    }
```

Comment:
Since there are not unique cases so we won't need the next_unique_minstrobe anyways and hence no false, but I'm not sure how to change it to void since the while loop depends on it.

Reply:
Well, do we need the while loop then?

Reply:
Perfect, resolved. (:+1: